src/Take.Blip.Client/Receivers/ContactMessageReceiverBase.cs

```diff
@@ -20,62 +20,76 @@ public abstract class ContactMessageReceiverBase : IMessageReceiver, IDisposable
         private readonly bool _cacheLocally;
         private readonly MemoryCache _contactCache;
         private readonly TimeSpan _cacheExpiration;
+        private readonly TimeSpan _getContactTimeout;
 
         /// <summary>
         /// Initializes a new instance of <see cref="ContactMessageReceiverBase"/> class.
         /// </summary>
         /// <param name="contactExtension"></param>
         /// <param name="directoryExtension"></param>
+        /// <param name="getContactTimeout"></param>
         /// <param name="cacheLocally">Indicates if the contact information should be cached locally.</param>
         /// <param name="cacheExpiration">Defines the local cache expiration, if configured.</param>
         protected ContactMessageReceiverBase(
             IContactExtension contactExtension,
             IDirectoryExtension directoryExtension,
             bool cacheLocally = true,
-            TimeSpan cacheExpiration = default(TimeSpan))
+            TimeSpan cacheExpiration = default,
+            TimeSpan getContactTimeout = default)
         {
             _directoryExtension = directoryExtension;
             _contactExtension = contactExtension;
             _cacheLocally = cacheLocally;
             _contactCache = new MemoryCache(new MemoryCacheOptions());
-            _cacheExpiration = cacheExpiration == default(TimeSpan) ? TimeSpan.FromMinutes(30) : cacheExpiration;
+            _cacheExpiration = cacheExpiration == default ? TimeSpan.FromMinutes(30) : cacheExpiration;
+            _getContactTimeout = getContactTimeout == default ? TimeSpan.FromSeconds(10) : getContactTimeout;
         }
 
         public async Task ReceiveAsync(Message envelope, CancellationToken cancellationToken = default(CancellationToken))
         {
             var identity = envelope.From.ToIdentity();
-
-            // First, tries get it from the cache, if configured.
             Contact contact = null;
-            if (_cacheLocally) contact = _contactCache.Get(identity.ToString()) as Contact;
-            
-            if (contact == null)
+            try
             {
-                try
+                using (var cts = new CancellationTokenSource(_getContactTimeout))
```

Comment:
You must link to the other cancellationToken

```suggestion
                using (var cts = new CancellationTokenSource(_getContactTimeout))
                using (var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cts.Token, cancellationToken))
```